quota/redis/redistb/redistb_test.go

```diff
@@ -0,0 +1,853 @@
+
+			argTimeSec, argTimeUs := timeToRedisPair(TestBaseTime)
+
+			resp := updateTokenBucketScript.Eval(
+				rdb,
+				keys.AsSlice(),
+
+				// Args
+				TestReplenishRate,
+				TestCapacity, // smaller than above
+				test.Requested,
+				argTimeSec,
+				argTimeUs,
+				"true",
+			)
+			assertRedisResults(t, resp,
+				true,
+				test.Expected,
+				argTimeSec,
+				argTimeUs,
+			)
+		})
+	}
+}
+
+func TestErrorIfMicrosecondsTooLarge(t *testing.T) {
+	rdb := redis.NewClient(&redis.Options{
+		Addr: "localhost:6379",
+	})
+
+	keys := makeKeys()
+	resp := updateTokenBucketScript.Eval(
+		rdb,
+		keys.AsSlice(),
+
+		// Args
+		TestReplenishRate,
+		TestCapacity,
+		1,
+		TestBaseTimeSec,
+		MicrosecondsInSecond,
+		"true",
+	)
+
+	err := resp.Err()
+	if err == nil {
+		t.Fatalf("expected an error, but got none")
+	}
+	if err.Error() != `now_us must be smaller than 10^6 (microseconds in a second)` {
+		t.Errorf("invalid error message: %s", err.Error())
+	}
+}
+
+const alphanumeric = "abcdefghijklmnopqrstuvwxyz" +
+	"ABCDEFGHIJKLMNOPQRSTUVWXYZ" +
+	"0123456789"
+
+func stringWithCharset(length int, charset string) string {
+	setlen := big.NewInt(int64(len(charset)))
+
+	b := make([]byte, length)
+	for i := range b {
+		ch, err := rand.Int(rand.Reader, setlen)
+		if err != nil {
+			panic(err)
+		}
+
+		b[i] = charset[ch.Int64()]
+	}
+	return string(b)
+}
+
+// Wrapper type for the keys we use in our call to the script.
+type redisKeys struct {
+	Prefix      string
+	Tokens      string
+	Refreshed   string
+	RefreshedUs string
+}
+
+// Get the keys in a slice format for passing to Eval
+func (r redisKeys) AsSlice() []string {
+	return []string{r.Tokens, r.Refreshed, r.RefreshedUs}
+}
+
+// Helper function to create Redis key names wrapper type
+func makeKeys() redisKeys {
+	var ret redisKeys
+
+	// Ensure each test ends up under a different key
+	ret.Prefix = stringWithCharset(20, alphanumeric)
+
+	// Use same key generation function as the live code
+	keys := tokenBucketKeys(ret.Prefix)
+	ret.Tokens = keys[0]
+	ret.Refreshed = keys[1]
+	ret.RefreshedUs = keys[2]
+	return ret
+}
+
+// Helper function to initialize Redis keys to given values
+func mustInitKeys(t *testing.T, c *redis.Client, keys redisKeys, tokens int64, refreshed time.Time) {
+	refreshedSec, refreshedUs := timeToRedisPair(refreshed)
+
+	if err := c.Set(keys.Tokens, tokens, TTL*time.Second).Err(); err != nil {
+		t.Fatalf("failed to set initial tokens: %v", err)
+	}
+	if err := c.Set(keys.Refreshed, refreshedSec, TTL*time.Second).Err(); err != nil {
+		t.Fatalf("failed to set refreshed time: %v", err)
+	}
+	if err := c.Set(keys.RefreshedUs, refreshedUs, TTL*time.Second).Err(); err != nil {
+		t.Fatalf("failed to set refreshed time (us): %v", err)
+	}
+}
+
+// Helper function that deserializes the returned values from our Redis script.
+func deserializeRedisResults(t *testing.T, resp *redis.Cmd) (bool, int64, int64, int64) {
+	results, err := resp.Result()
```

Comment:
Could call `t.Helper()` here (and in other helpers/asserts), if appropriate?

Reply:
Ooh, TIL - thank you! Done 👍